### PR TITLE
Bot friendly fire reduction

### PR DESCRIPTION
**Pull request by NickAragua**
Doesn't happen very often, but I've seen the bot blow up buildings containing its own turrets to try to get to the other side. Mostly this happens after you've considerably reduced the building CF. This PR addresses that by preventing the long-range pathfinder from pathing through buildings containing immobile units. 

Initial version was abominably slow for some reason, but adding a caching mechanism speeded things up nicely.

### CODE REVIEW

File: megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java [hunk 1]
```diff
@@ -42,6 +44,7 @@
 
     Comparator<BulldozerMovePath> movePathComparator;
     int maximumCost = Integer.MAX_VALUE;
+    Map<Coords, Boolean> friendlyFireCheckResults = new HashMap<>();
```
Comment (Windchild292) [inline megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java:47]: Access modifiers?

File: megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java [hunk 2]
```diff
@@ -230,6 +233,12 @@ protected void processChild(BulldozerMovePath child, List<BulldozerMovePath> chi
             return;
         }
         
+        // let's avoid pathing through buildings containing our immobile units - 
+        // they're not going to get out of the way and we can probably do better than killing our own guys
+        if(!child.isJumping() && friendlyFireCheck(child.getEntity(), child.getGame(), child.getFinalCoords(), false)) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java:238]:
```suggestion
        if (!child.isJumping() && friendlyFireCheck(child.getEntity(), child.getGame(), child.getFinalCoords(), false)) {
```

File: megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java [hunk 3]
```diff
@@ -242,6 +251,37 @@ protected void processChild(BulldozerMovePath child, List<BulldozerMovePath> chi
...
+        if (building == null) {
+            friendlyFireCheckResults.put(position, false);
+            return false;
+        }
+        
+        // check if there are any entities in the building that meet the following criteria:
+        // - is friendly
+        // - if we care only about mobile units, has no MP 
+        for(Entity entity : game.getEntitiesVector(position, true)) {
+            if(!entity.isEnemyOf(shooter) && (includeMobileUnits || (entity.getWalkMP(true, false) == 0))) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java:275]:
```suggestion
        for (Entity entity : game.getEntitiesVector(position, true)) {
            if (!entity.isEnemyOf(shooter) && (includeMobileUnits || (entity.getWalkMP(true, false) == 0))) {
```

File: megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java [hunk 4]
```diff
@@ -242,6 +251,37 @@ protected void processChild(BulldozerMovePath child, List<BulldozerMovePath> chi
         }
     }
     
+    /**
+     * Utility function that returns true if an attack on the building in the given coordinates
+     * will result in damage to friendly units. Computation is cached as it is somewhat expensive to perform for each possible path node.
+     */
+    public boolean friendlyFireCheck(Entity shooter, IGame game, Coords position, boolean includeMobileUnits) {
+        if(friendlyFireCheckResults.containsKey(position)) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java:259]:
```suggestion
        if (friendlyFireCheckResults.containsKey(position)) {
```

File: megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java [hunk 5]
```diff
@@ -242,6 +251,37 @@ protected void processChild(BulldozerMovePath child, List<BulldozerMovePath> chi
         }
     }
     
+    /**
+     * Utility function that returns true if an attack on the building in the given coordinates
+     * will result in damage to friendly units. Computation is cached as it is somewhat expensive to perform for each possible path node.
+     */
+    public boolean friendlyFireCheck(Entity shooter, IGame game, Coords position, boolean includeMobileUnits) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/pathfinder/DestructionAwareDestinationPathfinder.java:258]: Why public instead of private?

**Review by Windchild292 — APPROVED** [review]

**Comment by NickAragua** [comment]
Also included a minor "cleanup" of Server.java - it was outputting a bunch of unnecessary log entries in "gravity enhanced" environments.

**Comment by Windchild292** [comment]
Recommend swapping the server logger to the new error(this, string) format but otherwise it looks good

**Comment by NickAragua** [comment]
Think I'll skip it for now, since the underlying method just calls getClass() anyway.

**Pull request merged**